### PR TITLE
Creates a "busy" animation for players

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
Little indicator above a player when they're currently doing something.

<details>
<summary>vids</summary>

Perspective: You are the moth
dreamseeker_b2LA4PpPAr

Hides under runechat
dreamseeker_ZgkCWTGqDz

</details>

Todo:

- [x] Feedback?
- [x] Sneaky params so it doesn't spoil your stealth run
- [x] Possible refactor
- [x] Probably missed some "sneaky" actions
- [x] coggers
## Why It's Good For The Game

<details>
<summary>sound on:</summary>

https://github.com/tgstation/tgstation/assets/42397676/ad71c567-0202-4158-ba50-c2946375f988

</details>

## Changelog
:cl: jlsnow301, infraredbaron
add: Added a new UI element over players that are interacting, building, etc.
/:cl:

### CODE REVIEW

File: code/datums/progressbar.dm [hunk 1]
```diff
@@ -76,16 +83,17 @@
 		LAZYREMOVEASSOC(user.progressbars, bar_loc, src)
 		user = null
 
-	if(user_client)
+	if(user_client)		
 		clean_user_client()
 
 	bar_loc = null
 	bar = null
+	cog = null
```
Comment (LemonInTheDark) [inline code/datums/progressbar.dm:91]: if this is destroy we need to QDEL_NULL the cog

File: code/datums/progressbar.dm [hunk 2]
```diff
@@ -94,39 +102,59 @@
...
 	add_prog_bar_image_to_client()
 
 
-///Adds a smoothly-appearing progress bar image to the player's screen.
+/// Adds a smoothly-appearing progress bar image to the player's screen.
 /datum/progressbar/proc/add_prog_bar_image_to_client()
 	bar.pixel_y = 0
 	bar.alpha = 0
-	user_client.images += bar
+	user_client.images.Add(bar)
```
Comment (LemonInTheDark) [inline code/datums/progressbar.dm:133]: why?

File: code/datums/progressbar.dm [hunk 3]
```diff
@@ -2,39 +2,44 @@
...
-	///The type of our last value for bar_loc, for debugging
+	/// The type of our last value for bar_loc, for debugging
 	var/location_type
-	///Where to draw the progress bar above the icon
+	/// Where to draw the progress bar above the icon
 	var/offset_y
+	/// The progress visible to other players
+	var/obj/effect/overlay/vis/cog
+	/// The blank image to hide the cog
+	var/image/blank
```
Comment (LemonInTheDark) [inline code/datums/progressbar.dm:26]: dumb idea just dropped, couldn't we use the bar image to override the cog? or am I bein dumb

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/datums/progressbar.dm (hunk 3 above)
Comment (jlsnow301) [reply code/datums/progressbar.dm:26]: correct! this is what i was thinking earlier. we would need to invert it, sort of

File: code/datums/progressbar.dm (hunk 2 above)
Comment (jlsnow301) [reply code/datums/progressbar.dm:133]: idk

**Comment by Elli-Skala** [comment]
Worry about any progress bars that are supost to be for stealthy interactions/fully unseen

**Comment by the-orange-cow** [comment]
> Worry about any progress bars that are supost to be for stealthy interactions/fully unseen

Will this out heretics draining an influence?

**Comment by jlsnow301** [comment]
yall need to check the bullet points :/

File: code/__HELPERS/mobs.dm [hunk 4]
```diff
@@ -230,8 +230,12 @@ GLOBAL_LIST_EMPTY(species_list)
 		delay *= user.cached_multiplicative_actions_slowdown
 
 	var/datum/progressbar/progbar
+	var/datum/cogbar/cog
+
 	if(progress)
 		progbar = new(user, delay, target || user)
+		if(visibility == DO_PUBLIC && delay >= 1 SECONDS)
+			cog = new(user)
```
Comment (Jacquerel) [inline code/__HELPERS/mobs.dm:238]: Seems like you've located this code such that if we merge your _other_ PR then if NPC is performing a do_after, it won't display the cog. 
This seems undesirable to me, surely if we want an action to be visible we want it to be visible if it is being performed autonomously too?

**Review by Jacquerel — CHANGES_REQUESTED** [review]

File: code/__DEFINES/_flags.dm [hunk 5]
```diff
@@ -312,3 +312,9 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 #define EMOTE_VISIBLE (1<<1)
 /// Is it an emote that should be shown regardless of blindness/deafness
 #define EMOTE_IMPORTANT (1<<2)
+
+// Flags for cogbars/do_after visibility
+/// Is the action visible to others
+#define DO_PUBLIC (1<<0)
+/// Is the action hidden from others
+#define DO_HIDDEN (1<<1)
```
Comment (ZephyrTFA) [inline code/__DEFINES/_flags.dm:320]: these shouldnt be bitflags if they are exclusive

File: code/__HELPERS/mobs.dm (hunk 4 above)
Comment (Jacquerel) [reply code/__HELPERS/mobs.dm:238]: That's what the "hidden" flag is for, no? 
If it should be invisible when an NPC is doing it why would we want it to be visible if a player-controlled one is doing it?

Comment (jlsnow301) [reply code/__HELPERS/mobs.dm:241]: good point. with the other PR, i will format it to only not create the progbar.

**Comment by moocowswag** [comment]
the amount of times someone has shoved me when i was doing a do_after to build a blast door.....

finally there will be no benefit of doubt when it comes to if they knew what they were doing.

**Comment by ophaq** [comment]
This is really good. While we are at it, lets make it so that when you are moved while doing a medical procedure with a knife or saw, you have a chance of letting it loose on an unsuspecting passerby and slicing them up.

**Comment by moocowswag** [comment]
> This is really good. While we are at it, lets make it so that when you are moved while doing a medical procedure with a knife or saw, you have a chance of letting it loose on an unsuspecting passerby and slicing them up.

It would actually make more sense for you to accidently slash the person you are doing surgery on but it would give you a reason to escalate against the person that moved you.

**Comment by SomeoneYouProbablyKnow** [comment]
Finally I can move doctors doing surgery with greater efficiency

**Review by Wallemations — APPROVED** [review]

**Comment by Wallemations** [comment]
Actually, just making sure, this is below the runechat layer right?

**Comment by YakumoChen** [comment]
type baiting: part 2: fiddling with stuff edition

regardless, make the cogs brass

**Comment by EricZilla** [comment]
Could we get this for fishing? So often people try to talk to me while I'm in the minigame

**Comment by jlsnow301** [comment]
> Could we get this for fishing? So often people try to talk to me while I'm in the minigame

this is just in do_after now but i can see about another pr where this is added

File: code/datums/cogbar.dm [hunk 6]
```diff
@@ -0,0 +1,87 @@
+#define COGBAR_ANIMATION_TIME 5
```
Comment (san7890) [inline code/datums/cogbar.dm:1]:
```suggestion
#define COGBAR_ANIMATION_TIME 5 DECISECONDS
```

let's be clear

**Review by san7890 — CHANGES_REQUESTED** [review]

File: code/datums/cogbar.dm (hunk 6 above)
Comment (jlsnow301) [reply code/datums/cogbar.dm]: nope! just copy pasted over from progbars

**Pull request merged**